### PR TITLE
Gui/Part: improve gated single-pick fallback and face selection

### LINKED ISSUE #19243: Edge selection when it is coplanar with a face
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Problem description

It is not possible to select an edge (or a snap point of an edge), when it is coplanar with another face (for instance edge of the wall, which is connected to another wall), when non-transparent DrawStyles are selected.

See behavior in the video (only with wireframe mode is the user able to select the edge).

https://github.com/user-attachments/assets/562935ac-6aa2-4d80-9061-5b07abe06088

Sample file from the video:

[FC-edge-selection.FCStd.zip](https://github.com/user-attachments/files/18547425/FC-edge-selection.FCStd.zip)

The behavior was confirmed by another user in this forum thread: https://forum.freecad.org/viewtopic.php?t=94147

I did a test with Arch_Walls and also with Part_Boxes with the same result.

### Full version info

```shell
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.1.0dev.39980 (Git) Conda
Build type: Release
Branch: main
Hash: 836ad91f3ddd3ec63b84b8e3f87a767d5465261d
Python 3.11.11, Qt 5.15.15, Coin 4.0.3, Vtk 9.3.0, OCC 7.8.1
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Light.qss/FreeCAD Light/Fusion
```

### Subproject(s) affected?

Core

### Anything else?

Based on experience with different 3d modelling software, I expected to be able to select described edge even in non-transparent draw styles. I think enabling this would enhance user workflow.

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tritao**
Follow-up to #29705.

This PR improves gated single-pick behavior without changing ordinary ungated picking.

## Summary

There are two related changes here:

1. When the normal single-pick path is blocked by the active selection filter and the current pick list contains no allowed candidate, do one wider fallback pick.
2. For `Part` face-only filters, allow raw edge/vertex hits to resolve to the visible adjacent face instead of relying on Coin’s raw hit ordering.

This keeps the generic selection logic in `Gui` and the shape-specific face decision logic in `Part`.

## Behavior

The single-pick flow is now:

1. Use the normal picked-point list.
2. If an active selection filter rejects all candidates in that list, do a second wider single-pick search.
3. If that wider pass finds allowed candidates, choose the nearest allowed fallback hit.
4. Otherwise keep the existing blocked-selection behavior.

For `Part` face-only filters:

1. If the raw hit is already a face, keep it.
2. If the raw hit is an edge or vertex, gather adjacent candidate faces from topology.
3. Choose the visible adjacent face from rendered geometry within the pick neighborhood.
4. If no visible adjacent face can be resolved, keep the existing blocked-selection behavior.

The widened fallback currently uses `2x` the current viewer pick radius.

## Scope

This is intentionally limited to selection-filter behavior.

It does not change ordinary picking when:
- there is no active selection filter, or
- the current pick list already contains a valid allowed candidate.

## Implementation Notes

- `Gui` now threads generic pick context through the `ViewProvider` single-pick path.
- `Part` uses that context to implement face-only recovery locally, where topology and rendered-face geometry are available.
- Blocked fallback ranking now uses nearest allowed hit rather than the first allowed Coin hit.
- Repeated clicks on an already selected face under a face-only gate no longer escalate to a blocked parent-object selection.

### CODE REVIEW

**Comment by tritao** [comment]
@FEA-eng @tarman3 Would you like to test this and make sure it works as you expect?

**Comment by tarman3** [comment]
Tested
Selection of Vertexes much easier now

**Comment by tarman3** [comment]
One note
Selection radius gate return to standard value only after selection blocked

On screenshot below
Farthest point **B** selected instead of nearest **A**
 
Screenshot_20260508_074246_1000_lossy


https://github.com/user-attachments/assets/61cb8f66-85bf-46d3-9644-b0f57f7a8352

**Comment by FEA-eng** [comment]
Thank you for preparing the follow-up PR so quickly. I was going to create a follow-up issue, but you've been faster ;-)

Perhaps the face selection filter could be enabled here too?

image

IMO, hovering over the face's edge should be sufficient.

My goal would be to show the blocked sign only if the cursor is really far from anything suitable for selection. Of course, it will be shown much more frequently for vertices and sometimes edges (apart from their intersections), but for faces, it could go away pretty much completely.

image

**Comment by tarman3** [comment]
I tried to compile this changes, but can not select anything in 3d view and FreeCAD just crash after few seconds

Could you please rebase
Maybe this is just a merge conflicts

Thank you

**Comment by tritao** [comment]
@tarman3 @FEA-eng Can you please check the latest iteration?

Now face selection filter should work better when selecting hover edges, getting this behavior actually a lot more difficult than it seems at first, from my testing it works, but will be good to get more testing, especially edge cases.

**Comment by tarman3** [comment]
Could you please rebase

**Comment by tritao** [comment]
> Could you please rebase

Sure, latest push rebased to latest.

**Comment by FEA-eng** [comment]
@tritao Thank you, it works very well now.

One minor thing I've noticed (perhaps irrelevant, but I report it just in case) is that double-clicking on a face results in the blocked selection sign being shown:

blocked sel

**Comment by tritao** [comment]
> @tritao Thank you, it works very well now.

Thanks for testing, and that behavior seems worth fixing, latest push addresses it, can you check?

**Comment by tarman3** [comment]
Looks good

**Comment by tritao** [comment]
Updated again, improved code structure so its cleaner.

**Comment by maxwxyz** [comment]
This actually also fixes https://github.com/FreeCAD/FreeCAD/issues/19243 @semhustej FYI
But you need to turn on the edge selection filter:


https://github.com/user-attachments/assets/8c29cd9b-8eaf-4706-9c89-89eb076872db

**Review by maxwxyz — APPROVED** [review]

**Comment by tritao** [comment]
> This actually also fixes #29914 @semhustej FYI But you need to turn on the edge selection filter:

Nice! After this PR lands I can take a look at the unfiltered edge selection for that case too.

**Comment by tritao** [comment]
@hyarion Since you reviewed my previous selection-related PR, maybe you'd like to take a look at this one?

**Review by FEA-eng — APPROVED** [review]
Tested, all good

**Comment by tritao** [comment]
Made one more push to improve code readability and add some more comments in some places.

**Comment by semhustej** [comment]
> This actually also fixes #19243 @semhustej FYI But you need to turn on the edge selection filter:

Nice! So now there is at least this workaround.

File: src/Gui/Selection/SoFCUnifiedSelection.cpp [hunk 1]
```diff
@@ -111,28 +114,187 @@ void printPreselectionInfo(
...
+        context.selectionPickContext.repick.viewportRegion = &action->getViewportRegion();
+        context.selectionPickContext.repick.pickRadius = action->getPickRadius();
+    }
+    if (action && action->getEvent()) {
+        context.normalizedCursorPosition = action->getEvent()->getNormalizedPosition(
+            action->getViewportRegion()
+        );
+        context.selectionPickContext.normalizedCursorPosition = &*context.normalizedCursorPosition;
+        context.eventPosition = action->getEvent()->getPosition();
+        context.selectionPickContext.repick.eventPosition = &*context.eventPosition;
```
Comment (chennes) [inline src/Gui/Selection/SoFCUnifiedSelection.cpp]: I have to admit that this is faintly terrifying. In order for it to work, you're depending on NRVO (not the C++17 guaranteed copy elision, which I don't think applies here) -- what happens when compiling with `-O0`, `/Od`, etc.? Does it work on all of our compilers?

Comment (tritao) [reply src/Gui/Selection/SoFCUnifiedSelection.cpp]: Good point, thanks! Indeed this was quite unsafe, I've changed it to something safer and simpler.

**Comment by tritao** [comment]
Rebased on top of latest main too.

**Comment by chennes** [comment]
I just built this and tested it against a large model (~20mb): when running in Part's face-only selection mode, preselection with this PR is giving noticeably laggy results compared to baseline main. Have you done any performance analysis of this mechanism @tritao?

**Comment by tritao** [comment]
> I just built this and tested it against a large model (~20mb): when running in Part's face-only selection mode, preselection with this PR is giving noticeably laggy results compared to baseline main. Have you done any performance analysis of this mechanism @tritao?

I haven't, I will try to run at least some analysis and get some numbers, and see if there are any improvements to be had.

**Comment by tritao** [comment]
I've been taking a look at this and I don't think it makes sense trying to optimize it further at this time, it's just a by-product of the existing sub-par picking performance. To make actual progress on this, we need to make progress first on optimizing Coin GPU picking and improving the selection APIs to allow picking multiple entities at once.

It should be quite feasible to implement but first we need to make progress on integrating the new Coin renderer.

I'll keep this a draft for now and we can revisit at a later stage.
